abm.py

```diff
@@ -113,16 +114,14 @@ def wrapper(*args: Any, **kwargs: Any) -> Optional[T]:
     return collect_when_decorator
 
 
-def write_jsonlines(history: History, filename: str) -> None:
-    with open(filename, "w") as file:
-        for step in history:
-            file.write(json.dumps(step, default=str) + "\n")
+def write_jsonlines(history: History, file: TextIO) -> None:
+    for step in history:
```

Comment:
Changing `filename: str` to `file: TextIO` is fine with me (though will need to be changed downstream when analysing results in any notebooks).

Another way we could do this is just replace `open(...)` in each of `write_jsonlines` and `read_jsonlines` with `smart_open.open(...)` so that the input is still a `filename` and the API is the same?
But perhaps you are envisioning when working with it in the container this is a better API choice?

Reply:
Makes for a better API. `{write,read}_jsonlines` write and read JSONlines to a file-like object - they don't actually care what it is. They can then work with in-memory streams, files, `smart_open`, whatever.